File: C++/프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

/*
    스택과 큐 동시에 가동한다고 가정하면,
    prices 배열을 돌아. stack에 넣게 될 텐데, top 에 있는 값보다 크면 push, 
    작으면 pop. 근데 여기서 값을 넣는게 아니라 인덱스를 넣게되면, answer에 넣는 것도 수월해지지 않을까? 싶긴함.
*/

vector<int> solution(vector<int> prices) {
    vector<int> answer(prices.size());
    stack<int> s; // index를 넣기
    
    for (int i = 0; i < prices.size(); i++) {
        
        while (!s.empty() && prices[s.top()] > prices[i]) {
            int nowIdx = s.top();
            s.pop();
            answer[nowIdx] = i - nowIdx;
        }
        s.push(i);
    }
    
    // 1 2 4 6 35 42 9 465 6 66 7 1100 52536    
    while (!s.empty()) {
        int now = s.top();
        answer[now] = prices.size() - 1 - now;
        s.pop();
    }
    
    return answer;
}

File: Java/프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    /*
        투포인터, 스택?
        
        생각 1: 스택 사용해서, 가격 인덱스를 넣고 그 스택의 peek의 값과 현 위치의 가격을 비교
        그럼 다음 인덱스에서 어떻게 비교를 할 것인가? 비교한 후 어떤 로직을 적용할 것인가? 
        
    */
    public int[] solution(int[] prices) {
        Stack<Integer> s = new Stack<>();
        int[] answer = new int[prices.length];
        
        s.push(0);
        for (int i = 1; i < prices.length; i++) {
            
            while (!s.isEmpty() && prices[s.peek()] > prices[i]) {
                answer[s.peek()] = i - s.peek();
                s.pop();
            }
            s.push(i);
        }
        
        while (!s.isEmpty()) {
            answer[s.peek()] = prices.length - 1 - s.peek();
            s.pop();
        }
        
        return answer;
    }
}